File: commands/digitalungdom/verify.js
/* global db include abs_path client cooldown */

const validateObjectID = require( 'mongodb' ).ObjectID.isValid;
const ObjectID = require( 'mongodb' ).ObjectID;
const fs = require( 'fs-extra' );
const util = require( 'util' );
const readFile = util.promisify( fs.readFile );
const Hogan = require( 'hogan.js' );
const checkDiscordID = include( 'models/check' ).checkDiscordID;

const sendMail = include( 'utils/sendMail' ).sendMail;

module.exports = {
  name: 'verify',
  description: 'Koppla ditt Discord-konto till ditt DU konto.',
  aliases: [ 'verifera' ],
  group: 'digitalungdom',
  usage: 'verify <username|code>',
  example: 'verify kelszo',
  serverOnly: false,
  adminOnly: false,
  async execute( message, args ) {
    if ( message.channel.type === 'text' && !message.deleted ) message.delete();
    if ( args.length === 0 ) return message.author.send( 'Du måste skicka det användarnamn som du vill koppla till.' );
    const content = args[ 0 ];
    const discordID = message.author.id;

    const guild = await client.guilds.get( process.env.GUILD_ID );
    const user = await guild.fetchMember( message.author );
    if ( user.roles.find( r => r.name === 'verified' ) ) return message.author.send( 'Du är redan verifierad.' );
    if ( !( await checkDiscordID( message.author.id ) ).valid ) return message.author.send( 'Detta konto är redan kopplat till ett Digital Ungdom konto.' );

    if ( validateObjectID( content ) ) {
      const exists = ( await db.collection( 'users' ).findOneAndUpdate( { 'discordVerification': ObjectID( content ) }, {
        $set: { 'connectedApps.discord': discordID },
        $unset: { 'discordVerification': 1 },
      }, { projection: { '_id': 0, 'details.email': 1 } } ) ).value;
      if ( exists ) {
        const email = exists.details.email;

        const templateData = await readFile( abs_path( 'emails/verify/verifyConfirmation.mustache' ), 'utf8' );
        const template = Hogan.compile( templateData );
        const body = template.render( { name: message.author.username } );

        sendMail( email, 'Ditt Discord-konto är nu kopplat', body );

        const role = await guild.roles.find( r => r.name === 'verified' );
        const user = await guild.fetchMember( message.author );
        user.addRole( role );

        return message.author.send( 'Ditt konto är nu kopllat, grattis!' );
      } else message.author.send( 'Det finns inget konto kopplat till den id.' );
    } else {
      if ( new Date() < cooldown.verify[ message.author.id ] ) return message.author.send( 'Du får bara skicka ett verifikations mail per dag.' );
      const discordVerification = ObjectID();
      const search = { 'details.username': content.toLowerCase(), 'connectedApps.discord': { $exists: false } };
      const set = { $set: { 'discordVerification': discordVerification } };
      const projection = { projection: { '_id': 0, 'details.email': 1 }, collation: { locale: 'en', strength: 2 } };

      const exists = ( await db.collection( 'users' ).findOneAndUpdate( search, set, projection ) ).value;
      if ( exists ) {
        const email = exists.details.email;

        const templateData = await readFile( abs_path( 'emails/verify/verify.mustache' ), 'utf8' );
        const template = Hogan.compile( templateData );
        const body = template.render( { token: discordVerification } );

        let tomorrow = new Date();
        cooldown.verify[ message.author.id ] = tomorrow.setDate( tomorrow.getDate() + 1 );

        sendMail( email, 'Koppla ditt Discord-konto', body );
        return message.author.send( 'Ett email har skickats till kontot kopplat till det användarnamnet!' );
      } else {
        return message.author.send( 'Det gick inte att hitta ett konto, detta kan bero på tre saker:\n**1.** Det finns inget konto med det användarnamn.\n**2.** Du har redan skickat ett email, testa igen om 24 timmar.\n**3.** Du är redan verifierad.' );
      }
    }
  },
};

File: commands/digitalungdom/verify.go
package digitalungdom

import (
	"github.com/asaskevich/govalidator"
	"github.com/digitalungdom-se/dub/internal"
	"github.com/digitalungdom-se/dub/pkg"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/options"
)

var Verify = internal.Command{
	Name:        "verify",
	Description: "Koppla ditt discord konto till ditt Digital Ungdom konto",
	Aliases:     []string{"verifiera"},
	Group:       "digitalungdom",
	Usage:       "verify <username|email|token>",
	Example:     "verify kelvin",
	ServerOnly:  false,
	AdminOnly:   false,

	Execute: func(ctx *pkg.Context, server *internal.Server) error {
		msg, err := ctx.Reply("**BEARBETAR...**")
		if err != nil {
			return err
		}

		if !ctx.IsDM() {
			err := ctx.Delete()
			if err != nil {
				return err
			}
		}

		for _, role := range server.Guild.Roles {
			if role.Name == "verified" {
				verifiedID := role.ID
				for _, member := range server.Guild.Members {
					if member.User.ID == ctx.Message.Author.ID {
						if pkg.StringInSlice(verifiedID, member.Roles) {
							_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Du är redan verifierad.")
							return err
						}
					}
				}
			}
		}

		if len(ctx.Args) == 0 {
			_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Du måste ange ditt användarnamn eller email.")
			return err
		}

		queryOptions := new(options.FindOneAndUpdateOptions)

		var returnDocument options.ReturnDocument = options.After
		queryOptions.ReturnDocument = &returnDocument

		queryOptions.Projection = bson.M{
			"details.email": true,
		}

		if len(ctx.Args[0]) == 16 {
			email, err := server.Database.ConnectDUAccount(ctx.Args[0], ctx.Message.Author.ID)
			if err != nil || email == "" {
				_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Inget konto kunde hittas")

				return err
			}

			err = server.Mailer.SendVerifyDiscordConfirmation(email, ctx.Message.Author.Username)
			if err != nil {
				return err
			}

			var verifiedID string

			for _, role := range server.Guild.Roles {
				if role.Name == "verified" {
					verifiedID = role.ID
				}
			}

			err = ctx.Discord.GuildMemberRoleAdd(server.Guild.ID, ctx.Message.Author.ID, verifiedID)
			if err != nil {
				return err
			}

			_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Grattis ditt Discord konto är nu kopplat till Digital Ungdom!")

			return err
		}

		var user internal.User

		if govalidator.IsEmail(ctx.Args[0]) {
			user, err = server.Database.GetUserByEmail(ctx.Args[0])

			if err != nil {
				_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Inget konto kunde hittas")
				return err
			}

			if user.ConnectedApps.Discord != "" {
				_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Ditt Digital Ungdom kontot är redan kopplat till ett Discord konto.")
				return err
			}

		} else {
			user, err = server.Database.GetUserByUsername(ctx.Args[0])
			if err != nil {
				_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Inget konto kunde hittas")
				return err
			}

			if user.ConnectedApps.Discord != "" {
				_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Ditt Digital Ungdom kontot är redan kopplat till ett Discord konto.")
				return err
			}
		}

		token, err := server.Database.AddDiscordVerificationToken(user.ID.Hex())
		if err != nil {
			return err
		}

		err = server.Mailer.SendVerifyDiscord(user.Details.Email, token)
		if err != nil {
			return err
		}

		_, err = ctx.Discord.ChannelMessageEdit(msg.ChannelID, msg.ID, "Ett email har nu skickats till dig, följ instruktionerna där för att komma vidare.")
		if err != nil {
			return err
		}

		return nil
	},
}
